lib/guardian/post_guardian.rb

```diff
@@ -303,4 +306,36 @@ def can_unhide?(post)
   def can_skip_bump?
     is_staff? || @user.has_trust_level?(TrustLevel[4])
   end
+
+  private
+
+  def can_create_post_in_topic?(topic)
+    return false if !SiteSetting.enable_system_message_replies? && topic.try(:subtype) == "system_message"
+
+    (!SpamRule::AutoSilence.prevent_posting?(@user) || (!!topic.try(:private_message?) && topic.allowed_users.include?(@user))) && (
+      !topic ||
+      !topic.category ||
+      Category.post_create_allowed(self).where(id: topic.category.id).count == 1
+    )
+  end
+
+  def topic_memoize_key(topic)
+    # Milliseconds precision on Topic#updated_at so that we don't use memoized results after topic has been updated.
+    "#{topic.id}-#{(topic.updated_at.to_f * 1000).to_i}"
+  end
+
+  def can_see_post_topic?(post)
+    topic = post.topic
+
+    return can_see_topic?(topic) if !topic
```

Comment:
I had to read this a few times to understand what's happening.

`can_see_topic?` always returns `false` for nil values, and I don't see any reason that would ever change. So I'm tempted to suggest:

```suggestion
    return false if topic.nil?
```
which I think describes the intention much more clearly?

Reply:
That was something I tried initially but when topic is nil, we actually need to return true here otherwise a bunch of tests would break. Unfortunately, properly fixing this isn’t straight forward so I wanted to wait until the release is over before working on it again.

Reply:
> we actually need to return true here 

Oh... that's weird because `can_see_topic?(nil)` returns false per

https://github.com/discourse/discourse/blob/01af8c452d0ecb18e962036903e279c33fea9732/lib/guardian/topic_guardian.rb#L261-L262

Where does the `true` come from at the moment? 🤔

Reply:
O sorry I thought this was the can_create_post method XD